File: python/find_eventual_safe_states.py
class Solution:
    def eventualSafeNodes(self, graph: list[list[int]]) -> list[int]:
        """time complexity: O(E+V), space complexity: O(V), where E is the number of edges and V is the number of vertices (nodes)"""
        result: list[int] = []
        safe: dict[int, bool] = {}
        for i in range(len(graph)):
            if self.dfs(i, safe, graph):
                result.append(i)
        return result

    def dfs(self, node: int, safe: dict[int, bool], graph: list[list[int]]) -> bool:
        if node in safe:
            return safe[node]
        safe[node] = False
        for neighbor in graph[node]:
            if not self.dfs(neighbor, safe, graph):
                return False
        safe[node] = True
        return True


def test_eventualSafeNodes_case_1():
    # arrange
    graph: list[list[int]] = [[1, 2], [2, 3], [5], [0], [5], [], []]
    expected: list[int] = [2, 4, 5, 6]

    # act
    solution = Solution()
    actual = solution.eventualSafeNodes(graph)

    # assert
    assert expected == actual


def test_eventualSafeNodes_case_2():
    # arrange
    graph: list[list[int]] = [[1, 2, 3, 4], [1, 2], [3, 4], [0, 4], []]
    expected: list[int] = [4]

    # act
    solution = Solution()
    actual = solution.eventualSafeNodes(graph)

    # assert
    assert expected == actual

File: rust/src/code/find_eventual_safe_states.rs
struct Solution;

use std::collections::HashMap;

impl Solution {
    // NOTE: time complexity O(E + V), where E is the number of edges and V is the number of vertices
    pub fn eventual_safe_nodes(graph: Vec<Vec<i32>>) -> Vec<i32> {
        let mut safe: HashMap<i32, bool> = HashMap::new();
        let mut result: Vec<i32> = Vec::new();
        for i in 0..graph.len() {
            if Self::dfs(i as i32, &mut safe, &graph) {
                result.push(i as i32);
            }
        }
        result
    }

    fn dfs(node: i32, safe: &mut HashMap<i32, bool>, graph: &Vec<Vec<i32>>) -> bool {
        if let Some(&n) = safe.get(&node) {
            return n;
        } else {
            safe.insert(node, false);
            for neighbor in graph[node as usize].iter() {
                if !Self::dfs(*neighbor, safe, graph) {
                    return false;
                }
            }
            safe.entry(node).and_modify(|x| *x = true);
            return true;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_eventual_safe_nodes_case_1() {
        // arrange
        let graph: Vec<Vec<i32>> = vec![
            vec![1, 2],
            vec![2, 3],
            vec![5],
            vec![0],
            vec![5],
            vec![],
            vec![],
        ];
        let expected: Vec<i32> = vec![2, 4, 5, 6];

        // act
        let actual = Solution::eventual_safe_nodes(graph);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_eventual_safe_nodes_case_2() {
        // arrange
        let graph: Vec<Vec<i32>> =
            vec![vec![1, 2, 3, 4], vec![1, 2], vec![3, 4], vec![0, 4], vec![]];
        let expected: Vec<i32> = vec![4];

        // act
        let actual = Solution::eventual_safe_nodes(graph);

        // assert
        assert_eq!(expected, actual);
    }
}
